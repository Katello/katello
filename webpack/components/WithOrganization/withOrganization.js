/* eslint-disable react/jsx-indent */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import { translate as __ } from 'foremanReact/common/I18n';
import { get } from 'lodash';
import SetOrganization from '../SelectOrg/SetOrganization';
import Header from '../../containers/Application/Headers';
import * as organizationActions from '../../scenes/Organizations/OrganizationActions';

const mapStateToProps = state => ({
  organization: state.katello.organization,
});

const mapDispatchToProps = dispatch => bindActionCreators({ ...organizationActions }, dispatch);

function withOrganization(WrappedComponent) {
  class CheckOrg extends Component {
    constructor(props) {
      super(props);
      this.state = { orgId: null };
    }
    static getDerivedStateFromProps(newProps, state) {
      const orgNodeId = document.getElementById('organization-id').dataset.id;

      if (state.orgId !== orgNodeId) {
        return { orgId: orgNodeId };
      }
      return null;
    }

    componentDidUpdate(prevProps) {
      const { location } = this.props;
      const orgTitle = get(location, 'state.orgChanged');
      const prevOrgTitle = get(prevProps, 'location.state.orgChanged');

      if (orgTitle !== prevOrgTitle) {
        window.tfm.nav.changeOrganization(orgTitle);
      }
    }

    render() {
      const { organization, location } = this.props;
      const newOrgSelected = get(location, 'state.orgChanged');

      if (newOrgSelected) {
        if (!organization.label && !organization.loading) { this.props.loadOrganization(); }
        return <WrappedComponent {...this.props} />;
      } else if (this.state.orgId === '') {
        return (
          <>
            <Header title={__('Select Organization')} />
            <SetOrganization />
          </>);
      }
      return <WrappedComponent {...this.props} />;
    }
  }

  CheckOrg.propTypes = {
    location: PropTypes.shape({}),
    loadOrganization: PropTypes.func.isRequired,
    organization: PropTypes.shape({
      label: PropTypes.string,
      loading: PropTypes.bool,
    }).isRequired,
  };

  CheckOrg.defaultProps = {
    location: undefined,
  };

  return connect(mapStateToProps, mapDispatchToProps)(CheckOrg);
}

export default withOrganization;
