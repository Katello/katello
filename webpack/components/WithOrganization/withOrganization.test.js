import React from 'react';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import withOrganization from './withOrganization';

jest.mock('../SelectOrg/SetOrganization');
const mockStore = configureMockStore([thunk]);
const store = mockStore({ katello: { organization: {} } });

describe('subscriptions page', () => {
  const WrappedComponent = () => <div> Wrapped! </div>;

  it('should render the wrapped component', () => {
    global.document.getElementById = () => ({ dataset: { id: 1 } });

    const Component = withOrganization(WrappedComponent);
    const page = mount(<Component store={store} />);
    expect(toJson(page)).toMatchSnapshot();
  });

  it('should render select org page', () => {
    global.document.getElementById = () => ({ dataset: { id: '' } });

    const Component = withOrganization(WrappedComponent);
    const page = mount(<Component store={store} />);
    expect(toJson(page)).toMatchSnapshot();
  });
});
