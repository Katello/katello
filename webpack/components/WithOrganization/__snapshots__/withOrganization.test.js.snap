// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`subscriptions page should render select org page 1`] = `
<Connect(CheckOrg)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <CheckOrg
    default={[Function]}
    loadOrganization={[Function]}
    organization={Object {}}
    saveOrganization={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
  >
    <Header
      title="Select Organization"
    >
      <HelmetWrapper
        defer={true}
        encodeSpecialCharacters={true}
      >
        <SideEffect(NullComponent)
          defer={true}
          encodeSpecialCharacters={true}
          title="Select Organization"
          titleAttributes={Object {}}
        >
          <NullComponent
            defer={true}
            encodeSpecialCharacters={true}
            title="Select Organization"
            titleAttributes={Object {}}
          />
        </SideEffect(NullComponent)>
      </HelmetWrapper>
    </Header>
    <Connect(withRouter(SetOrganization)) />
  </CheckOrg>
</Connect(CheckOrg)>
`;

exports[`subscriptions page should render the wrapped component 1`] = `
<Connect(CheckOrg)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <CheckOrg
    default={[Function]}
    loadOrganization={[Function]}
    organization={Object {}}
    saveOrganization={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
  >
    <WrappedComponent
      default={[Function]}
      loadOrganization={[Function]}
      organization={Object {}}
      saveOrganization={[Function]}
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
      storeSubscription={
        Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        }
      }
    >
      <div>
         Wrapped! 
      </div>
    </WrappedComponent>
  </CheckOrg>
</Connect(CheckOrg)>
`;
